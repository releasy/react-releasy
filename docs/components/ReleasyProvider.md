# ReleasyProvider

The root component.

## Usage

First create an instance of `Config`:

```javascript
import { Config, Link } from 'react-releasy';

const link = new Link({
  url: 'https://yourserveraddress.com/graphql',
});

const config = new Config({
  link,
});
```

Then wrap the application:

```javascript
import { ReleasyProvider } from 'react-releasy';

ReactDOM.render(
  <ReleasyProvider config={config}>
    <MyApplication />
  </ReleasyProvider>,
  document.getElementById('root'),
);
```

After this we can inject the `Relay` `environment` into our components using [withReleasy](../hocs/withReleasy.md), or even getting it directly:

```javascript
import { getEnvironment } from 'react-releasy';


const environment = getEnvironment();
```

## Props

- **config**: see [Config](../classes/Config.md)

----

[Go back](../)
