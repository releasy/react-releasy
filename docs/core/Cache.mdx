---
name: Cache
menu: Core
route: /core/cache
---

# Cache

A class responsible to store/recover data without needing make any request to the server.

## Usage

```javascript
import { InMemoryCache } from 'react-releasy';

const cache = new InMemoryCache({
  ttl: 300000,
  size: 250,
});
```
