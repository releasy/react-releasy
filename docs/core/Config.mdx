---
name: Config
menu: Core
route: /core/config
---

# Config

A configuration class.

## Usage

```javascript
import { Config, InMemoryCache, Link } from 'react-releasy';

const config = new Config({
  link: new Link({ url: 'https://yourserveraddress.com/graphql' }),
  cache: new InMemoryCache(),
  devTools: true,
  networkLogger: true,
});
```
