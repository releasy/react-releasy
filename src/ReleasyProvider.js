// @flow
import * as React from 'react';
import * as PropTypes from 'prop-types';
import { Environment } from 'relay-runtime';
import invariant from 'invariant';

import { getEnvironment } from './store';

type Props = {
  children: React.Node,
};

class ReleasyProvider extends React.PureComponent<Props> {
  static propTypes = {
    children: PropTypes.element.isRequired,
  };

  static childContextTypes = {
    environment: PropTypes.instanceOf(Environment).isRequired,
  };

  constructor(props: Props) {
    invariant(props.children, 'ReleasyProvider: missing "children"');

    super(props);
  }

  getChildContext() {
    return {
      environment: getEnvironment(),
    };
  }

  render() {
    return this.props.children;
  }
}

export default ReleasyProvider;
