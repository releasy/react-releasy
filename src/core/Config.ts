import { CacheConfig, RequestNode, UploadableMap, Variables, GraphQLResponse } from 'relay-runtime';
import invariant from 'invariant';

export interface LinkInterface {
  fetch: (
    request: RequestNode,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables: UploadableMap,
  ) => Promise<GraphQLResponse>;
}

export interface CacheInterface {
  get: (queryID: string, variables: Variables) => GraphQLResponse;
  set: (queryID: string, variables: Variables, payload: GraphQLResponse) => void;
  clear: () => void;
  size: () => number;
}

type ConfigPropsType = {
  link: LinkInterface,
  cache?: CacheInterface,
  devTools?: boolean,
  networkLogger?: boolean,
};

class Config {
  link: LinkInterface;
  cache: CacheInterface;
  devTools: boolean;
  networkLogger: boolean;

  constructor(props: ConfigPropsType) {
    invariant(props, 'Config: missing props');
    invariant(props.link, 'Config: missing "link"');

    this.link = props.link;
    this.cache = props.cache || null;
    this.devTools = props.devTools || false;
    this.networkLogger = props.networkLogger || false;
  }
}

export default Config;
