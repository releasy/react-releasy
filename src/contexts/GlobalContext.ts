import * as React from 'react';
import { Environment } from 'relay-runtime';

import Config from '../core/Config';

export type GlobalContextValueType = {
  config: Config,
  environment: Environment,
};

const value: GlobalContextValueType = {
  config: null,
  environment: null,
};

export const getConfig = (): Config => value.config;

export const getEnvironment = (): Environment => value.environment;

export const setConfig = (config: Config): void => {
  value.config = config;
};

export const setEnvironment = (environment: Environment): void => {
  value.environment = environment;
};

const { Provider, Consumer } = React.createContext<GlobalContextValueType>(value);

export default {
  Provider,
  Consumer,
};
