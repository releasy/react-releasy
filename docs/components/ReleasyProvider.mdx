---
name: ReleasyProvider
menu: Components
route: /components/releasy-provider
---

import { PropsTable } from 'docz'
import { ReleasyProvider } from '../../src'

# ReleasyProvider

## Properties

<PropsTable of={ReleasyProvider} />