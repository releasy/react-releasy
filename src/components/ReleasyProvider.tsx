import * as React from 'react';
import invariant from 'invariant';

import GlobalContext, {
  GlobalContextValueType,
  setConfig,
  setEnvironment,
} from '../contexts/GlobalContext';
import Config from '../core/Config';
import createEnvironment from '../relay/createEnvironment';

type ReleasyProviderPropsType = {
  config: Config,
  children: React.ReactNode,
};

type ReleasyProviderStateType = {
  globalContextValue: GlobalContextValueType,
};

class ReleasyProvider extends React.PureComponent<
  ReleasyProviderPropsType,
  ReleasyProviderStateType
> {
  constructor(props: ReleasyProviderPropsType) {
    super(props);

    invariant(props.config, 'ReleasyProvider: missing "config"');

    const { config } = props;
    const environment = createEnvironment(config);

    setConfig(config);
    setEnvironment(environment);

    this.state = {
      globalContextValue: {
        config,
        environment,
      },
    };
  }

  render() {
    const { children } = this.props;
    const { globalContextValue } = this.state;

    return (
      <GlobalContext.Provider value={globalContextValue}>
        {children}
      </GlobalContext.Provider>
    );
  }
}

export default ReleasyProvider;
