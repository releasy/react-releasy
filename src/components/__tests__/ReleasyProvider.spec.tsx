import * as React from 'react';
import { shallow } from 'enzyme';

import { createConfig } from '../../../tests/fixtures/core';

import { getConfig, getEnvironment } from '../../contexts/GlobalContext';
import ReleasyProvider from '../ReleasyProvider';

it('should render a valid ReleasyProvider', () => {
  const config = createConfig();

  const component = shallow(
    <ReleasyProvider config={config}>
      <div />
    </ReleasyProvider>,
  );

  expect(component.instance()).toMatchSnapshot();
});

it('should be capable to get Config statically', () => {
  const config = createConfig();

  shallow(
    <ReleasyProvider config={config}>
      <div />
    </ReleasyProvider>,
  );

  expect(getConfig()).toMatchSnapshot();
});

it('should be capable to get Environment statically', () => {
  const config = createConfig();

  shallow(
    <ReleasyProvider config={config}>
      <div />
    </ReleasyProvider>,
  );

  expect(getEnvironment()).toMatchSnapshot();
});
