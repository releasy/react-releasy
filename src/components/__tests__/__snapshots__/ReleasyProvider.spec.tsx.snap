// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be capable to get Config statically 1`] = `
Config {
  "cache": null,
  "devTools": false,
  "link": Link {
    "fetch": [Function],
    "fetchTimeout": 30000,
    "headers": Object {},
    "retryDelays": Array [
      1000,
      3000,
      5000,
      10000,
    ],
    "url": "/graphql",
  },
  "networkLogger": false,
}
`;

exports[`should be capable to get Environment statically 1`] = `
RelayModernEnvironment {
  "__setNet": [Function],
  "_deferrableSelections": Set {},
  "_network": Object {
    "execute": [Function],
    "executeWithEvents": [Function],
  },
  "_publishQueue": RelayPublishQueue {
    "_appliedOptimisticUpdates": Set {},
    "_backup": Object {},
    "_handlerProvider": [Function],
    "_pendingBackupRebase": false,
    "_pendingData": Set {},
    "_pendingOptimisticUpdates": Set {},
    "_pendingUpdaters": Set {},
    "_store": RelayMarkSweepStore {
      "_gcEnabled": true,
      "_hasScheduledGC": false,
      "_index": 0,
      "_recordSource": Object {},
      "_roots": Map {},
      "_subscriptions": Set {},
      "_updatedRecordIDs": Object {},
    },
  },
  "_store": RelayMarkSweepStore {
    "_gcEnabled": true,
    "_hasScheduledGC": false,
    "_index": 0,
    "_recordSource": Object {},
    "_roots": Map {},
    "_subscriptions": Set {},
    "_updatedRecordIDs": Object {},
  },
  "configName": undefined,
  "unstable_internal": Object {
    "areEqualSelectors": [Function],
    "createFragmentSpecResolver": [Function],
    "createOperationSelector": [Function],
    "getDataIDsFromObject": [Function],
    "getFragment": [Function],
    "getRequest": [Function],
    "getSelector": [Function],
    "getSelectorList": [Function],
    "getSelectorsFromObject": [Function],
    "getVariablesFromObject": [Function],
  },
}
`;

exports[`should render a valid ReleasyProvider 1`] = `
ReleasyProvider {
  "context": Object {},
  "props": Object {
    "children": <div />,
    "config": Config {
      "cache": null,
      "devTools": false,
      "link": Link {
        "fetch": [Function],
        "fetchTimeout": 30000,
        "headers": Object {},
        "retryDelays": Array [
          1000,
          3000,
          5000,
          10000,
        ],
        "url": "/graphql",
      },
      "networkLogger": false,
    },
  },
  "refs": Object {},
  "setState": [Function],
  "state": Object {
    "globalContextValue": Object {
      "config": Config {
        "cache": null,
        "devTools": false,
        "link": Link {
          "fetch": [Function],
          "fetchTimeout": 30000,
          "headers": Object {},
          "retryDelays": Array [
            1000,
            3000,
            5000,
            10000,
          ],
          "url": "/graphql",
        },
        "networkLogger": false,
      },
      "environment": RelayModernEnvironment {
        "__setNet": [Function],
        "_deferrableSelections": Set {},
        "_network": Object {
          "execute": [Function],
          "executeWithEvents": [Function],
        },
        "_publishQueue": RelayPublishQueue {
          "_appliedOptimisticUpdates": Set {},
          "_backup": Object {},
          "_handlerProvider": [Function],
          "_pendingBackupRebase": false,
          "_pendingData": Set {},
          "_pendingOptimisticUpdates": Set {},
          "_pendingUpdaters": Set {},
          "_store": RelayMarkSweepStore {
            "_gcEnabled": true,
            "_hasScheduledGC": false,
            "_index": 0,
            "_recordSource": Object {},
            "_roots": Map {},
            "_subscriptions": Set {},
            "_updatedRecordIDs": Object {},
          },
        },
        "_store": RelayMarkSweepStore {
          "_gcEnabled": true,
          "_hasScheduledGC": false,
          "_index": 0,
          "_recordSource": Object {},
          "_roots": Map {},
          "_subscriptions": Set {},
          "_updatedRecordIDs": Object {},
        },
        "configName": undefined,
        "unstable_internal": Object {
          "areEqualSelectors": [Function],
          "createFragmentSpecResolver": [Function],
          "createOperationSelector": [Function],
          "getDataIDsFromObject": [Function],
          "getFragment": [Function],
          "getRequest": [Function],
          "getSelector": [Function],
          "getSelectorList": [Function],
          "getSelectorsFromObject": [Function],
          "getVariablesFromObject": [Function],
        },
      },
    },
  },
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_element": <ReleasyProvider
        config={
          Config {
            "cache": null,
            "devTools": false,
            "link": Link {
              "fetch": [Function],
              "fetchTimeout": 30000,
              "headers": Object {},
              "retryDelays": Array [
                1000,
                3000,
                5000,
                10000,
              ],
              "url": "/graphql",
            },
            "networkLogger": false,
          }
        }
      >
        <div />
      </ReleasyProvider>,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_newState": null,
      "_rendered": <Context.Provider
        value={
          Object {
            "config": Config {
              "cache": null,
              "devTools": false,
              "link": Link {
                "fetch": [Function],
                "fetchTimeout": 30000,
                "headers": Object {},
                "retryDelays": Array [
                  1000,
                  3000,
                  5000,
                  10000,
                ],
                "url": "/graphql",
              },
              "networkLogger": false,
            },
            "environment": RelayModernEnvironment {
              "__setNet": [Function],
              "_deferrableSelections": Set {},
              "_network": Object {
                "execute": [Function],
                "executeWithEvents": [Function],
              },
              "_publishQueue": RelayPublishQueue {
                "_appliedOptimisticUpdates": Set {},
                "_backup": Object {},
                "_handlerProvider": [Function],
                "_pendingBackupRebase": false,
                "_pendingData": Set {},
                "_pendingOptimisticUpdates": Set {},
                "_pendingUpdaters": Set {},
                "_store": RelayMarkSweepStore {
                  "_gcEnabled": true,
                  "_hasScheduledGC": false,
                  "_index": 0,
                  "_recordSource": Object {},
                  "_roots": Map {},
                  "_subscriptions": Set {},
                  "_updatedRecordIDs": Object {},
                },
              },
              "_store": RelayMarkSweepStore {
                "_gcEnabled": true,
                "_hasScheduledGC": false,
                "_index": 0,
                "_recordSource": Object {},
                "_roots": Map {},
                "_subscriptions": Set {},
                "_updatedRecordIDs": Object {},
              },
              "configName": undefined,
              "unstable_internal": Object {
                "areEqualSelectors": [Function],
                "createFragmentSpecResolver": [Function],
                "createOperationSelector": [Function],
                "getDataIDsFromObject": [Function],
                "getFragment": [Function],
                "getRequest": [Function],
                "getSelector": [Function],
                "getSelectorList": [Function],
                "getSelectorsFromObject": [Function],
                "getVariablesFromObject": [Function],
              },
            },
          }
        }
      >
        <div />
      </Context.Provider>,
      "_rendering": false,
      "_updater": [Circular],
    },
  },
  Symbol(enzyme.__setState__): [Function],
}
`;
